connectors/src/connectors/snowflake/temporal/activities.ts

```diff
@@ -1,7 +1,157 @@
+import logger from "@connectors/logger/logger";
+
+const snowflakeTableCodec = t.type({
+  name: t.string,
+  database_name: t.string,
+  schema_name: t.string,
+});
+
+export async function syncSnowflakeConnection(connectorId: ModelId) {
+  const getConnectorAndCredentialsRes = await getConnectorAndCredentials({
+    connectorId,
+    logger,
+  });
+  if (getConnectorAndCredentialsRes.isErr()) {
+    throw getConnectorAndCredentialsRes.error;
+  }
+
+  const { credentials, connector } = getConnectorAndCredentialsRes.value;
+
+  const tablesRes = await fetchTables({ credentials });
+  if (tablesRes.isErr()) {
+    throw tablesRes.error;
+  }
+
+  const tablesValidation = t.array(snowflakeTableCodec).decode(tablesRes.value);
+  if (isLeft(tablesValidation)) {
+    const pathError = reporter.formatValidationErrors(tablesValidation.left);
+    throw new Error(`Invalid tables response: ${pathError}`);
+  }
+  const [allDatabases, allSchemas, allTables] = await Promise.all([
+    RemoteDatabaseModel.findAll({
+      where: {
+        connectorId,
+      },
+    }),
+    RemoteSchemaModel.findAll({
+      where: {
+        connectorId,
+      },
+    }),
+    RemoteTableModel.findAll({
+      where: {
+        connectorId,
+      },
+    }),
+  ]);
+
+  const readGrantedInternalIds = new Set([
+    ...allDatabases.map((db) => db.internalId),
+    ...allSchemas.map((s) => s.internalId),
+    ...allTables
+      .filter((t) => t.permission === "selected")
+      .map((t) => t.internalId),
+  ]);
+  const tableByInternalId = Object.fromEntries(
+    allTables.map((table) => [table.internalId, table])
+  );
+
+  const parseTableInternalId = (
+    tableInternalId: string
+  ): [string, string, string] => {
+    const [dbName, schemaName, tableName] = tableInternalId.split(".");
+    if (!dbName || !schemaName || !tableName) {
+      throw new Error(`Invalid table internalId: ${tableInternalId}`);
+    }
+
+    return [dbName, schemaName, tableName];
+  };
+
+  const isTableReadGranted = (tableInternalId: string) => {
+    const [dbName, schemaName] = parseTableInternalId(tableInternalId);
+
+    const schemaInternalId = [dbName, schemaName].join(".");
+
+    return (
+      readGrantedInternalIds.has(dbName) ||
+      readGrantedInternalIds.has(schemaInternalId) ||
+      readGrantedInternalIds.has(tableInternalId)
+    );
+  };
+
+  const tablesOnSnowflake = tablesValidation.right;
+  const internalIdsOnSnowflake = new Set(
+    tablesOnSnowflake.map(
+      (t) => `${t.database_name}.${t.schema_name}.${t.name}`
+    )
+  );
+
+  for (const internalId of internalIdsOnSnowflake) {
+    if (isTableReadGranted(internalId)) {
+      let table = tableByInternalId[internalId];
+      if (!table || !table.lastUpsertedAt) {
+        if (!table) {
+          const [dbName, schemaName, tableName] =
+            parseTableInternalId(internalId);
+          table = await RemoteTableModel.create({
+            connectorId,
+            internalId,
```

Comment:
nit: I was confused a bit wether internalId refer to the snowflakeInternalid or not (looking at the code, it's the same thing).

Reply:
Yeah `internalId` is a name convention we use in `connectors` -- agree it's a bit confusing